src/drivers/magnetometer/qmc5883p/QMC5883P.cpp

```diff
@@ -0,0 +1,308 @@
+			}
+		}
+	}
+
+	return PX4_ERROR;
+}
+
+void QMC5883P::RunImpl()
+{
+    const hrt_abstime now = hrt_absolute_time();
+
+    switch (_state) {
+    case STATE::RESET:
+        {
+	// CNTL2: Software Reset
+        RegisterWrite(Register::CNTL2, CNTL2_BIT::SOFT_RST);
+	_reset_timestamp = now;
+        _failure_count = 0;
+        _state = STATE::WAIT_FOR_RESET;
+        perf_count(_reset_perf);
+        ScheduleDelayed(100_ms); // POR Completion Time
+        break;
+	}
+    case STATE::WAIT_FOR_RESET:
+	{
+        	// SOFT_RST: This bit is automatically reset to zero after POR routine
+		if ((RegisterRead(Register::CHIP_ID) == Chip_ID)
+		    && ((RegisterRead(Register::CNTL2) & CNTL2_BIT::SOFT_RST) == 0)) {
+
+			// if reset succeeded then configure
+			_state = STATE::CONFIGURE;
+			ScheduleDelayed(10_ms);
+
+		} else {
+			// RESET not complete
+			if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				PX4_DEBUG("Reset failed, retrying");
+				_state = STATE::RESET;
+				ScheduleDelayed(100_ms);
+
+			} else {
+				PX4_DEBUG("Reset not complete, check again in 10 ms");
+				ScheduleDelayed(10_ms);
+			}
+		}
+
+		break;
+	}
+    case STATE::CONFIGURE:
+        if (Configure()) {
+            // if configure succeeded then start reading every 20 ms (50 Hz)
+            _state = STATE::READ;
+            ScheduleOnInterval(20_ms, 20_ms);
+
+        } else {
+                // CONFIGURE not complete
+			    if (hrt_elapsed_time(&_reset_timestamp) > 1000_ms) {
+				    PX4_DEBUG("Configure failed, resetting");
+				    _state = STATE::RESET;
+
+			    } else {
+				    PX4_DEBUG("Configure failed, retrying");
+			    }
+
+            ScheduleDelayed(100_ms);
+        }
+
+        break;
+
+    case STATE::READ: {
+                    struct TransferBuffer{
+                        uint8_t X_LSB;
+				        uint8_t X_MSB;
+				        uint8_t Y_LSB;
+				        uint8_t Y_MSB;
+				        uint8_t Z_LSB;
+				        uint8_t Z_MSB;
+				        uint8_t STATUS;
+                    }buffer{};
+
+                	bool success = false;
+			        uint8_t cmd = static_cast<uint8_t>(Register::X_LSB);
+
+                    if (transfer(&cmd, 1, (uint8_t *)&buffer, sizeof(buffer)) == PX4_OK) {
+                    // process data if successful transfer, no overflow
+				    if ((buffer.STATUS & STATUS_BIT::OVL) == 0) {
+					    int16_t x = combine(buffer.X_MSB, buffer.X_LSB);
+					    int16_t y = combine(buffer.Y_MSB, buffer.Y_LSB);
+					    int16_t z = combine(buffer.Z_MSB, buffer.Z_LSB);
+
+					if (x != _prev_data[0] || y != _prev_data[1] || z != _prev_data[2]) {
+						_prev_data[0] = x;
+						_prev_data[1] = y;
+						_prev_data[2] = z;
+
+						// Sensor orientation(Amovlab_F410)
+						//  Forward X := Z
+						//  Right   Y := Y
+						//  Down    Z := -X
+						int16_t new_x = (z == INT16_MIN) ? INT16_MAX : z;
+						int16_t new_y = (y == INT16_MIN) ? INT16_MAX : y;
+						int16_t new_z = (x == INT16_MIN) ? INT16_MAX : -x;
```

Comment:
This doesn't match the datasheet. PX4 is FRD:

| PX4    | QMC |
| -------- | ------- |
| F  | -y    |
| R | -x     |
| D    | -z    |


![image](https://github.com/user-attachments/assets/3993e3fa-68c5-4ca4-967a-2f312f7857d3)